Review followup: PR #23440 — t3588: surface gh upgrade remediation

## Unaddressed review bot suggestions

PR #23440 was merged with unaddressed review bot feedback. Each comment
below includes its file path, line number, a direct link to the inline
review comment, and a `diff` fence with the code context the bot was
flagging. Resolved and outdated threads are filtered out via GitHub's
GraphQL review-thread state. Read the relevant lines, decide whether
the suggestion is correct, and either apply the fix or close this issue
with a wontfix rationale.

**Source PR:** https://github.com/marcusquinn/aidevops/pull/23440

---

### You are the triager (worker-is-triager rule)

This issue is **auto-created from review bot output** and dispatched
directly to you. Review bots can be wrong: hallucinated line refs, false
premises about codebase structure, template-driven sweeps without
measurements (see GH#17832-17835 for prior art and `AGENTS.md`
"AI-Generated Issue Quality"). **Do not assume the bot is correct.** Verify before acting.

You must end in exactly one of three outcomes — no fourth "hand it back
to the human" path exists. Humans approve decisions; they do not re-do
analysis.

#### Outcome A — Premise falsified → close the issue

1. Read the cited `file:line` (listed under *Files to modify* below).
2. If the bot's claim is factually wrong (file doesn't exist at that
   line, function doesn't behave as described, "auto-generated" section
   isn't actually auto-generated, etc.), **close the issue** with a
   comment in this shape:

   > **Premise falsified.** \<what the bot claimed\>. \<what the code
   > actually shows, with a `file:line` citation or one-line quote\>.
   > Not acting.

   No PR. No further dispatch. The closing comment trains the next
   session reading this thread and the noise filter.

#### Outcome B — Premise correct + fix is obvious → implement and PR

1. Verify the bot's premise as above.
2. Read the Worker Guidance section below, open a worktree, implement.
3. Open a PR with `Resolves #<this-issue-number>` in the body
   (use THIS issue's number, not the source PR's) so merge auto-closes it.
4. Follow the normal Lifecycle Gate (brief, tests, review-bot-gate,
   merge, postflight).

#### Outcome C — Premise correct but approach is a genuine judgment call

Only use this path if you reach it after Outcomes A and B don't apply:
the bot's finding is real, but the fix requires a decision that is
architectural, policy, breaking-change, or otherwise genuinely outside
what you can resolve autonomously. In that case, post a **decision
comment** with exactly these fields:

- **Premise check:** one line, confirming the finding is real.
- **Analysis:** 2-4 bullets on the trade-offs.
- **Recommended path:** the option you would take if the decision were
  yours, with rationale.
- **Specific question:** the single decision the human needs to make
  (yes/no or pick-one, not open-ended).

Then apply `needs-maintainer-review` and stop. The human wakes up to a
ready-to-approve recommendation, not a blank task.

> **Ambiguity about scope or style is not Outcome C.** Per
> `AGENTS.md` "Reasoning responsibility", the model does the
> thinking and delivers a recommendation. Only escalate what is genuinely
> a maintainer-only decision.

### Worker Guidance

**Files to modify:**

- `.agents/scripts/setup/modules/tool-install.sh:102`
- `.agents/scripts/setup/modules/tool-install.sh:61`

**Implementation steps (Outcome B path):**

1. Read the `diff` block under each inline comment below — it shows the
   exact code the bot was flagging. Open the file only if you need
   surrounding context beyond what the diff tail shows.
2. Read the bot's full comment below the diff — it contains the rationale
   and any suggested change.
3. Verify the premise before implementing (see Outcome A). If the premise
   is wrong, switch to Outcome A instead of burning iterations trying to
   satisfy a wrong suggestion.
4. If multiple comments target the same file, group your edits into one
   logical commit.
5. Run `shellcheck` / `markdownlint-cli2` / project tests as appropriate.

**Verification:**

- Open the new PR with `Resolves #<this-issue>` so this followup is auto-closed on merge.
- If the bot's suggestion was incorrect, close this issue with a Outcome A comment — do not open a no-op PR.

### Inline comments

#### gemini-code-assist on `.agents/scripts/setup/modules/tool-install.sh:61`

[View inline comment](https://github.com/marcusquinn/aidevops/pull/23440#discussion_r3223152371)

```diff
+				_print_gh_slurp_manual_upgrade
+			fi
+		else
+			print_warning "Package-manager gh upgrade failed or was unavailable"
+			_print_gh_slurp_manual_upgrade
+		fi
+	else
+		print_info "Skipped GitHub CLI upgrade"
+		_print_gh_slurp_manual_upgrade
+	fi
+	return 0
+}
```

> ![medium](https://www.gstatic.com/codereviewagent/medium-priority.svg)
> 
> The `_offer_gh_slurp_upgrade` function should return a success/failure status instead of always returning 0, allowing the caller to react to the result. This aligns with the repository rule to propagate exit codes and avoid masking errors. The use of `if command; then` is preferred over checking `$?` directly, which aligns with ShellCheck SC2181.
> 
> ```shell
> _offer_gh_slurp_upgrade() {
> 	local pkg_manager="$1"
> 	local os_name=""
> 	os_name=$(uname -s 2>/dev/null || printf 'unknown')
> 
> 	if [[ "$os_name" != "Linux" ]]; then
> 		_print_gh_slurp_manual_upgrade
> 		return 1
> 	fi
> 
> 	echo ""
> 	print_warning "Linux GitHub CLI is below the aidevops minimum. Old distro packages can break pulse dispatch."
> 	if [[ "$pkg_manager" == "unknown" ]]; then
> 		print_warning "No supported package manager detected for an automatic gh upgrade attempt"
> 		_print_gh_slurp_manual_upgrade
> 		return 1
> 	fi
> 	setup_prompt upgrade_gh_cli "Try to upgrade GitHub CLI (gh) using ${pkg_manager}? [y/N]: " "N"
> 	# shellcheck disable=SC2154  # set indirectly by setup_prompt via read
> 	if [[ "$upgrade_gh_cli" =~ ^[Yy]$ ]]; then
> 		print_info "Attempting to upgrade gh using ${pkg_manager}..."
> 		if install_packages "$pkg_manager" gh; then
> 			if declare -F aidevops_gh_slurp_supported >/dev/null 2>&1 && aidevops_gh_slurp_supported; then
> 				print_success "GitHub CLI now satisfies the aidevops prerequisite"
> 				return 0
> 			else
> 				print_warning "gh still does not satisfy the aidevops prerequisite after package-manager upgrade"
> 				_print_gh_slurp_manual_upgrade
> 			fi
> 		else
> 			print_warning "Package-manager gh upgrade failed or was unavailable"
> 			_print_gh_slurp_manual_upgrade
> 		fi
> 	else
> 		print_info "Skipped GitHub CLI upgrade"
> 		_print_gh_slurp_manual_upgrade
> 	fi
> 	return 1
> }
> ```
> 
> <details>
> <summary>References</summary>
> 
> 1. In shell scripts, functions that wrap a command should propagate its exit code to the caller. Avoid masking potential errors with a har

#### gemini-code-assist on `.agents/scripts/setup/modules/tool-install.sh:102`

[View inline comment](https://github.com/marcusquinn/aidevops/pull/23440#discussion_r3223152372)

```diff
@@ -45,14 +94,18 @@ setup_git_clis() {
 		print_success "Found Git CLI tools: ${cli_tools[*]}"
 	fi
 
+	local pkg_manager
+	pkg_manager=$(detect_package_manager)
+
+	if [[ "$gh_needs_slurp_upgrade" == "true" ]]; then
+		_offer_gh_slurp_upgrade "$pkg_manager"
+	fi
```

> ![medium](https://www.gstatic.com/codereviewagent/medium-priority.svg)
> 
> Update the caller to check the return status of `_offer_gh_slurp_upgrade` and update the `gh_needs_slurp_upgrade` variable accordingly. This ensures the final success message is only displayed if the prerequisite is actually met.
> 
> ```suggestion
> 	if [[ "$gh_needs_slurp_upgrade" == "true" ]]; then
> 		if _offer_gh_slurp_upgrade "$pkg_manager"; then
> 			gh_needs_slurp_upgrade="false"
> 		fi
> 	fi
> ```
> 
> <details>
> <summary>References</summary>
> 
> 1. In shell scripts, capture a command's exit code in a variable instead of using $? directly in conditionals. This aligns with ShellCheck SC2181 and improves clarity for multi-way branches.
> </details>

### PR review summaries

#### gemini-code-assist[bot] review summary

[View review](https://github.com/marcusquinn/aidevops/pull/23440#pullrequestreview-4268536893)

> ## Code Review
> 
> This pull request introduces version checks and remediation logic for the GitHub CLI (gh), specifically requiring version 2.51.0 or higher to support the --slurp flag. The changes include new warning messages, manual upgrade guidance, and an interactive upgrade prompt within the setup script for Linux environments. Additionally, new test suites in both JavaScript and Bash have been added to verify warning classification and remediation workflows. Feedback suggests improving error propagation in the shell scripts by having the upgrade function return a success/failure status and updating the caller to handle this status correctly.
<!-- aidevops:sig -->
---
[aidevops.sh](https://aidevops.sh) v3.15.34 automated scan.

--- a/.agents/scripts/setup/modules/tool-install.sh
+++ b/.agents/scripts/setup/modules/tool-install.sh
@@ -28,7 +28,7 @@ _offer_gh_slurp_upgrade() {
 
 	if [[ "$os_name" != "Linux" ]]; then
 		_print_gh_slurp_manual_upgrade
-		return 0
+		return 1
 	fi
 
 	echo ""
@@ -36,7 +36,7 @@ _offer_gh_slurp_upgrade() {
 	if [[ "$pkg_manager" == "unknown" ]]; then
 		print_warning "No supported package manager detected for an automatic gh upgrade attempt"
 		_print_gh_slurp_manual_upgrade
-		return 0
+		return 1
 	fi
 	setup_prompt upgrade_gh_cli "Try to upgrade GitHub CLI (gh) using ${pkg_manager}? [y/N]: " "N"
 	# shellcheck disable=SC2154  # set indirectly by setup_prompt via read
@@ -45,6 +45,7 @@ _offer_gh_slurp_upgrade() {
 		if install_packages "$pkg_manager" gh; then
 			if declare -F aidevops_gh_slurp_supported >/dev/null 2>&1 && aidevops_gh_slurp_supported; then
 				print_success "GitHub CLI now satisfies the aidevops prerequisite"
+				return 0
 			else
 				print_warning "gh still does not satisfy the aidevops prerequisite after package-manager upgrade"
 				_print_gh_slurp_manual_upgrade
@@ -57,7 +58,7 @@ _offer_gh_slurp_upgrade() {
 		print_info "Skipped GitHub CLI upgrade"
 		_print_gh_slurp_manual_upgrade
 	fi
-	return 0
+	return 1
 }
 
 setup_git_clis() {
@@ -98,7 +99,9 @@ setup_git_clis() {
 	pkg_manager=$(detect_package_manager)
 
 	if [[ "$gh_needs_slurp_upgrade" == "true" ]]; then
-		_offer_gh_slurp_upgrade "$pkg_manager"
+		if _offer_gh_slurp_upgrade "$pkg_manager"; then
+			gh_needs_slurp_upgrade="false"
+		fi
 	fi
 
 	# Offer to install missing tools

--- a/.agents/scripts/tests/test-gh-slurp-remediation.sh
+++ b/.agents/scripts/tests/test-gh-slurp-remediation.sh
@@ -80,8 +80,8 @@ assert_contains() {
 
 make_fake_gh "gh version 2.50.0"
 old_notice=$(call_update_prereq_check)
-assert_contains "update-check emits warning for old gh" "$old_notice" "[WARN] GitHub CLI (gh) 2.50.0 is too old"
-assert_contains "update-check points Linux users at setup" "$old_notice" "Run aidevops setup to upgrade gh"
+assert_contains "update-check emits warning for old gh" "$old_notice" "GitHub CLI (gh) detected version 2.50.0 is too old"
+assert_contains "update-check points Linux users at setup" "$old_notice" "install or upgrade from the official GitHub CLI package repository"
 
 make_fake_gh "gh version 2.51.0"
 new_notice=$(call_update_prereq_check)
@@ -102,10 +102,14 @@ setup_offer_output=$(bash -c "
 	aidevops_gh_slurp_supported() { return 1; }
 	AIDEVOPS_GH_MIN_SLURP_VERSION=2.51.0
 	source '$TOOL_INSTALL_SCRIPT'
+	set +eE
+	trap - ERR
 	_offer_gh_slurp_upgrade apt
+	printf 'STATUS:%s\n' "\$?"
 ")
 assert_contains "setup offers Linux gh upgrade" "$setup_offer_output" "Try to upgrade GitHub CLI (gh) using apt?"
 assert_contains "setup skipped path prints manual guidance" "$setup_offer_output" "official GitHub CLI package source"
+assert_contains "setup skipped path returns failure" "$setup_offer_output" "STATUS:1"
 
 setup_yes_output=$(bash -c "
 	set +e
@@ -118,10 +122,40 @@ setup_yes_output=$(bash -c "
 	aidevops_gh_slurp_supported() { return 0; }
 	AIDEVOPS_GH_MIN_SLURP_VERSION=2.51.0
 	source '$TOOL_INSTALL_SCRIPT'
+	set +eE
+	trap - ERR
 	_offer_gh_slurp_upgrade apt
+	printf 'STATUS:%s\n' "\$?"
 ")
 assert_contains "setup accepted path invokes package upgrade" "$setup_yes_output" "INSTALL:apt"
 assert_contains "setup accepted path rechecks prerequisite" "$setup_yes_output" "GitHub CLI now satisfies"
+assert_contains "setup accepted path returns success" "$setup_yes_output" "STATUS:0"
+
+setup_git_clis_output=$(bash -c "
+	set +e
+	test_bin=\$(mktemp -d)
+	trap 'rm -rf \"\$test_bin\"' EXIT
+	printf '#!/usr/bin/env bash\nexit 0\n' >\"\$test_bin/gh\"
+	printf '#!/usr/bin/env bash\nexit 0\n' >\"\$test_bin/glab\"
+	chmod +x \"\$test_bin/gh\" \"\$test_bin/glab\"
+	PATH=\"\$test_bin:\$PATH\"
+	print_info() { local msg=\"\$*\"; printf 'INFO:%s\n' \"\$msg\"; return 0; }
+	print_warning() { local msg=\"\$*\"; printf 'WARN:%s\n' \"\$msg\"; return 0; }
+	print_success() { local msg=\"\$*\"; printf 'OK:%s\n' \"\$msg\"; return 0; }
+	setup_prompt() { local var_name=\"\$1\" prompt_text=\"\$2\"; printf '%s' \"\$prompt_text\"; printf -v \"\$var_name\" '%s' 'Y'; return 0; }
+	detect_package_manager() { printf 'apt\n'; return 0; }
+	install_packages() { slurp_supported=true; printf 'INSTALL:%s\n' \"\$*\"; return 0; }
+	uname() { printf 'Linux\n'; return 0; }
+	slurp_supported=false
+	aidevops_gh_slurp_status_message() { printf 'GitHub CLI (gh) is too old\n'; return 0; }
+	aidevops_gh_slurp_supported() { [[ \"\$slurp_supported\" == 'true' ]]; return \$?; }
+	AIDEVOPS_GH_MIN_SLURP_VERSION=2.51.0
+	source '$TOOL_INSTALL_SCRIPT'
+	set +eE
+	trap - ERR
+	setup_git_clis
+")
+assert_contains "setup_git_clis clears gh remediation flag after successful upgrade" "$setup_git_clis_output" "All Git CLI tools installed and ready"
 
 printf '\n%d tests run, %d failed\n' "$TESTS_RUN" "$TESTS_FAILED"
 if [[ "$TESTS_FAILED" -gt 0 ]]; then

